larpandoracontent/LArControlFlow/TrackDirectionTool.cc

```diff
@@ -0,0 +1,1154 @@
+      const float slidingFitPitch(LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+
+      LArTrackStateVector trackStateVector;
+      LArPfoHelper::GetSlidingFitTrajectory(pPfo, pVertex, m_slidingFitWindow, slidingFitPitch, trackStateVector);
+
+     
+
+      const Cluster *const pClusterW = GetTargetClusterFromPFO(pPfo, trackStateVector);
+
+      CaloHitList totalcalohits;
+      LArPfoHelper::GetCaloHits(pPfo, TPC_VIEW_U, totalcalohits);
+      LArPfoHelper::GetCaloHits(pPfo, TPC_VIEW_V, totalcalohits);
+      LArPfoHelper::GetCaloHits(pPfo, TPC_VIEW_W, totalcalohits);
+
+
+        if (pClusterW->GetNCaloHits() <= m_minClusterCaloHits)
+        {
+            throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+        }
+
+        DirectionFitObject finalDirectionFitObject = GetClusterDirection(pClusterW);
+        this->ComputeProbability(finalDirectionFitObject);
+
+        //If the PFO is 3D, then 3D endpoints should be set 
+        if (LArPfoHelper::IsThreeD(pPfo))
+            SetEndpoints(finalDirectionFitObject, trackStateVector);
+
+ 
+        return finalDirectionFitObject;
+    }
+
+    catch (StatusCodeException &statusCodeException)
+    {
+        throw statusCodeException;
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::WriteLookupTableToTree(LookupTable &lookupTable)
+{
+    std::vector<int> mapVector1, reverseMapVector2;
+    std::vector<double> mapVector2, reverseMapVector1;
+
+    for (auto &element : lookupTable.GetMap())
+    {
+        mapVector1.push_back(element.first);
+        mapVector2.push_back(element.second);
+    }
+
+    for (auto &element : lookupTable.GetReverseMap())
+    {
+        reverseMapVector1.push_back(element.first);
+        reverseMapVector2.push_back(element.second);
+    }
+
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName.c_str(), "mapVector1", &mapVector1));
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName.c_str(), "mapVector2", &mapVector2));
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName.c_str(), "reverseMapVector1", &reverseMapVector1));
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName.c_str(), "reverseMapVector2", &reverseMapVector2));
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName.c_str(), "maxRange", lookupTable.GetMaxRange()));
+    PANDORA_MONITORING_API(FillTree(this->GetPandora(), m_treeName.c_str()));
+}
+
+  //-----------------------------------------------------------------------------------------------------------------------
+
+const Cluster* TrackDirectionTool::GetTargetClusterFromPFO(const ParticleFlowObject* pPfo, const LArTrackStateVector &trackStateVector)
+{
+    ClusterList clusterListW;
+    LArPfoHelper::GetTwoDClusterList(pPfo, clusterListW);
```

Comment:
Hum, I guess it would. But I think it should only work with actual W clusters. Will have a closer look at this (as this part of code was written by Joris).